### ISSUE TITLE
unexpected token

### ISSUE THREAD

**Issue by AngryBerryMS** [fvclaus/vsc-sort-json-array#14]
When I tried this extension with the following Json file, it encountered a `unexpected token` error.

![test2](https://user-images.githubusercontent.com/67643974/105117106-e7b0ff00-5b06-11eb-8ccd-1377100e1193.gif)

Based on [this answer](https://stackoverflow.com/questions/14432165/uncaught-syntaxerror-unexpected-token-with-json-parse), could you please handle when the case have some hidden/special characters?

```javascript
export default function parseArray(array: string, fileExtension: FileExtension) : unknown[] {
  let object;
  try {
    object = JSON.parse(applyPreFilter(array.replace(/\\n/g, "\\n")  
               .replace(/\\'/g, "\\'")
               .replace(/\\"/g, '\\"')
               .replace(/\\&/g, "\\&")
               .replace(/\\r/g, "\\r")
               .replace(/\\t/g, "\\t")
               .replace(/\\b/g, "\\b")
               .replace(/\\f/g, "\\f");, fileExtension));
  } catch (e) {
    throw new Error(`Cannot parse selection as JSON. Reason: ${e}`);
  }
  if (object.constructor === Array) {
    return object;
  } else {
    throw new Error(`Selection is a ${object.constructor} not an array.`);
  }
}
```

JSON file for test:

```json
{
    "redirections": [
      {
        "source_path": "a/about.md",
        "redirect_url": "/test/about",
        "redirect_document_id": false
      },
      {
        "source_path": "b/javascript.md",
        "redirect_url": "/test/js",
        "redirect_document_id": false
      },
      {
        "source_path": "b/javascript.md",
        "redirect_url": "/test/aaa",
        "redirect_document_id": false
      }
  ]
}
```

**Comment by fvclaus** [comment]
Thanks for creating an issue and providing a fix as well!

I am having problems understanding the issue. I read through the linked StackOverflow answer as well as all comments, but still don't understand it. There seems to be a problem with special characters, but the code does not actually replace special characters. Using `\\n` as an example:
- `\\n` is a literal `\` followed by an `n`, therefore printed as `\n`
- The node repl will "render" it as `\\n`, unless it is printed to `console.log`
```
> '\\n'
'\\n'
> console.log('\\n')
\n
> "\\n".replace(/\\n/, "\\n")
'\\n'
> console.log("\\n".replace(/\\n/, "\\n"))
\n
```
So the code is replacing a `\\n` with a `\\n`. That does not change anything.

Looking at your example, there seems to be a problem with whitespace, but I can't figure it out. Pasting your array to an vsc or any other editor shows me normal whitespaces and the array can be parsed successfully.

![vsc_example](https://user-images.githubusercontent.com/2384738/105625932-fb24d880-5e2c-11eb-957f-e01b7506892b.png)

Can you upload your file the example that is not working?

**Comment by AngryBerryMS** [comment]
Sorry for the late reply.

I found an easy way to resolve this probelm, just press `ctrl+shift+p` in vs code and enter `format`, and after formatting this extension can work properly.

**Comment by fvclaus** [comment]
Sure. That is a workaround. Do you still have the file in a state where the parsing fails? I would like to take a look at it. Others might have the same issue and I was planning to write  a more forgiving parser anyway.

**Comment by AngryBerryMS** [comment]
Yes. You can use this file to reproduce the problem (download and remove the `.txt` extension)

[openpublishing.redirection.test.json.txt](https://github.com/fvclaus/vsc-sort-json-array/files/5884113/openpublishing.redirection.test.json.txt)

**Comment by fvclaus** [comment]
Thanks for sharing the file. The white space before the last array element are not normal spaces but a non-breaking spaces.
See: https://en.wikipedia.org/wiki/Non-breaking_space

Non-breaking spaces are not valid white spaces in JSON (or any language I know). I will add code that can handle them, because that might be a common issue. I will leave the issue open until this is fixed as a reminder for me and for other people to see.

**Comment by fvclaus** [comment]
This is a reminder for myself: JS supports even more whitespace characters. A list can be found here: https://tc39.es/ecma262/#sec-white-space
Below are the line terminators that are supported. I should support them as well.

**Comment by fvclaus** [comment]
I have added support for more whitespace characters in 4.1.0. Your example above works now:

![non breaking whitespace fix](https://github.com/user-attachments/assets/f9c46c66-7201-488d-9e27-625de62b004f)
